src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractSyncNamespaceService.ChangeNamespaceCodeAction.cs

```diff
@@ -0,0 +1,482 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.AddImports;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.FindSymbols;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.RemoveUnnecessaryImports;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Simplification;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CodeRefactorings.SyncNamespace
+{
+    internal abstract partial class AbstractSyncNamespaceService<TNamespaceDeclarationSyntax, TCompilationUnitSyntax>
+        where TNamespaceDeclarationSyntax : SyntaxNode
+        where TCompilationUnitSyntax : SyntaxNode 
+    {
+        /// <summary>
+        /// This code action tries to change the name of the namespace declaration to 
+        /// match the folder hierarchy of the document. The new namespace is constructed 
+        /// by concatenate the default namespace of the project and all the folders in 
+        /// the file path up to the project root.
+        /// 
+        /// For example, if he default namespace is `A.B.C`, file path is 
+        /// "[project root dir]\D\E\F\Class1.cs" and declared namespace in the file is
+        /// `Foo.Bar.Baz`, then this action will change the namespace declaration
+        /// to `A.B.C.D.E.F`. 
+        /// 
+        /// Note that it also handles the case where the target namespace or declared namespace 
+        /// is global namespace, i.e. default namespace is "" and the file is located at project 
+        /// root directory, and no namespace declaration in the document, respectively.
+        /// </summary>
+        internal sealed class ChangeNamespaceCodeAction : CodeAction
```

Comment:
it's unclear to me why we need to put the code in the subclass.  This isn't how we normally do things . just subclass SolutionChangeAction and pass in an appropriate lambda to figure things out right...?